KeyPressedPreferNative: More performat key input

phu54321 opened the pull request:
On windows, `Gdx.input.isKeyPressed` uses lwjgl's event system for fetching key inputs. This event-based system sometimes lags due to OS issues, so beatoraja had problem getting key state ASAP. When pressing multiple keys, some keys were registered later than other keys.

This commit fixes the issue by utilizing Windows's `GetAsyncKeyState` function for fetching current keyboard state itself.

- Before: note that simultaneous keys are not registered simultaneously.
- After: simultaneous keys are registered simultaneously.

[![Comparison before & after](https://img.youtube.com/vi/L5pCyFWzfvo/0.jpg)](https://youtu.be/L5pCyFWzfvo)


Reference: previous libgdx input polling system (from LWJGL2)


Notes:

- ~~This adds jna to the dependencies.~~ Fixed. No additional dependencies.
-  *Whether the code works on mac/linux is not yet tested*. Testing is required.

https://github.com/LWJGL/lwjgl/blob/master/src/java/org/lwjgl/opengl/WindowsKeyboard.java

phu54321 committed:
KeyPressedPreferNative: More performat key input

On windows, `Gdx.input.isKeyPressed` uses lwjgl's event system
for fetching key inputs. This event-based system sometimes
lags due to OS issues, so beatoraja had problem getting key state
ASAP. When pressing multiple keys, some keys were registered
later than other keys.

This commit fixes the issue by utilizing Windows's `GetAsyncKeyState`
function for fetching current keyboard state itself. Note that this
function will poll key state regardless of whether the window is
focused or not, so this could be documented.

Reference: previous libgdx input polling system (from LWJGL2)

https://github.com/LWJGL/lwjgl/blob/master/src/java/org/lwjgl/opengl/WindowsKeyboard.java

phu54321 committed:
fix(KeyPressedPreferNative): disallow key input when the window is not focused

phu54321 commented:
Fix) beatoraja now only accept keys when the window is focused.

- GetAsyncKeyState works regardless of whether the window is focused.
- Added `Display.IsActive()` check so that the key will input only when the window is focused.

phu54321 committed:
perf(KeyPressedPreferNative): Use GetForegroundWindow instead of Display.isActive()

Display.isActive() requires a global lock on the display object, so it frequently stalls, resulting in less-than-optimal gaming experience.

To make things lock-free, we again use JNA to implement `Display.isActive()` ourselves.

Note that this code is coded for lwjgl2 backend the beatoraja currently uses. To maximize modularity, All input-specific code are
confined in KeyPressedPreferNative class only.

phu54321 committed:
fix(KeyPressedPreferNative): Fix Alt+F4

Alt+F4 should kill the application.

Merrg1n commented:
I had noticed this issue before, which led me to write this project https://github.com/Merrg1n/beatoraja-ime-fix. After reading the source code, I found that libGDX handles keyboard input events tied to the game loop. This means the actual keyboard polling rate is bound to the game's frame rate. However, the `GetAsyncKeyState` API only exists on Windows, and more consideration is needed for Mac/Linux (since this project is cross-platform).

phu54321 commented:
> This means the actual keyboard polling rate is bound to the game's frame rate.

https://github.com/exch-bms2/beatoraja/blob/91d975358619fb0cb9acc3b267897a2872d29ec5/src/bms/player/beatoraja/MainController.java#L363
Polling happens on a separate thread, so polling rate is much faster than the display refresh rate.

Separate consideration is required for macOS/Linux. Every windows-specific aspects are enclosed on the `KeyPressedPreferNative` class, and support for other platforms also can be encapsulated.

Merrg1n commented:
> Polling happens on a separate thread, so polling rate is much faster than the display refresh rate.

Let me add some implementation details about LibGDX. Even though it appears that keyboard events are polled by another thread in the game, this is actually how it works.

In LibGDX 1.9.9 (the version actually used in this repository), [`Gdx.input.isKeyPressed()`](https://github.com/libgdx/libgdx/blob/404ad6893f58fe880b91416df1a3af15eb3c6ed8/gdx/src/com/badlogic/gdx/Input.java#L657) actually calls [`com.badlogic.gdx.backends.lwjgl.LwjglInput.isKeyPressed()`](https://github.com/libgdx/libgdx/blob/404ad6893f58fe880b91416df1a3af15eb3c6ed8/backends/gdx-backend-lwjgl/src/com/badlogic/gdx/backends/lwjgl/LwjglInput.java#L227). (In later versions of LibGDX, this call path has changed.)
```java
        public boolean isKeyPressed (int key) {
		if (!Keyboard.isCreated()) return false;

		if (key == Input.Keys.ANY_KEY)
			return pressedKeys > 0;
		else
			return Keyboard.isKeyDown(getLwjglKeyCode(key));
	}
```
This method actually calls [`org.lwjgl.input.Keyboard.isKeyDown()`](https://github.com/LWJGL/lwjgl/blob/2df01dd762e20ca0871edb75daf670ccacc89b60/src/java/org/lwjgl/input/Keyboard.java#L403), and then attempts to retrieve a key from the `keyDownBuffer`.
```java
    public static boolean isKeyDown(int key) {
        synchronized(OpenGLPackageAccess.global_lock) {
            if (!created) {
                throw new IllegalStateException("Keyboard must be created before you can query key state");
            } else {
                return keyDownBuffer.get(key) != 0;
            }
        }
    }
```
So next, we should focus on where the `keyDownBuffer` is updated. It's actually updated in [`org.lwjgl.input.Keyboard.poll()`](https://github.com/LWJGL/lwjgl/blob/2df01dd762e20ca0871edb75daf670ccacc89b60/src/java/org/lwjgl/input/Keyboard.java#L383), This method calls [`org.lwjgl.opengl.InputImplementation.pollKeyboard()`](https://github.com/LWJGL/lwjgl/blob/2df01dd762e20ca0871edb75daf670ccacc89b60/src/java/org/lwjgl/opengl/InputImplementation.java#L115), which is implemented by platform-specific classes. On Windows, it's implemented in [`org.lwjgl.opengl.WindowsDisplay`](https://github.com/LWJGL/lwjgl/blob/2df01dd762e20ca0871edb75daf670ccacc89b60/src/java/org/lwjgl/opengl/WindowsDisplay.java#L704), and ultimately, [it calls `GetAsyncKeyState` API](https://github.com/LWJGL/lwjgl/blob/2df01dd762e20ca0871edb75daf670ccacc89b60/src/java/org/lwjgl/opengl/WindowsKeyboard.java#L68).
```java
    public static void poll() {
        synchronized(OpenGLPackageAccess.global_lock) {
            if (!created) {
                throw new IllegalStateException("Keyboard must be created before you can poll the device");
            } else {
                implementation.pollKeyboard(keyDownBuffer);
                read();
            }
        }
    }
```
It seems like it should work correctly, but the `Keyboard.poll()` method is actually called within [`org.lwjgl.opengl.Display.pollDevices()`](https://github.com/LWJGL/lwjgl/blob/2df01dd762e20ca0871edb75daf670ccacc89b60/src/java/org/lwjgl/opengl/Display.java#L670), which is then invoked by 
[`org.lwjgl.opengl.Display.processMessages()`](https://github.com/LWJGL/lwjgl/blob/2df01dd762e20ca0871edb75daf670ccacc89b60/src/java/org/lwjgl/opengl/Display.java#L595). Finally, this brings us back to LibGDX, where it is called in [`com.badlogic.gdx.backends.lwjgl.LwjglApplication.mainLoop()`](https://github.com/libgdx/libgdx/blob/404ad6893f58fe880b91416df1a3af15eb3c6ed8/backends/gdx-backend-lwjgl/src/com/badlogic/gdx/backends/lwjgl/LwjglApplication.java#L140), So, in reality, the keyboard event is polled within the main game loop.
```java
	void mainLoop () {
                // ...
		graphics.lastTime = System.nanoTime();
		boolean wasActive = true;
		while (running) {
			Display.processMessages();
                        // ...
                }
       }
```
`LwjglApplication.mainLoop()` includes the actual rendering loop, so the polling of keyboard events is tied to the game’s rendering loop. As a result, the actual keyboard polling rate is linked to the game's frame rate.

phu54321 commented:
Wow, awesome description :) I haven't noticed that behavior, but it seems very plausible. Probably I'll need to test it too!

My PR could also help mitigate that, as my code doesn't rely on the lwjgl's event loop system.

---

Minor corrections: On `WindowKeyboard.poll`, this code fills up the `keyDownBuffer`.
https://github.com/LWJGL/lwjgl/blob/2df01dd762e20ca0871edb75daf670ccacc89b60/src/java/org/lwjgl/opengl/WindowsKeyboard.java#L80C3-L80C38

`key_down_buffer` is filled by `handleKey` function, which is in turn called in response to `WM_KEYDOWN` and `WM_KEYUP` events.
https://github.com/LWJGL/lwjgl/blob/2df01dd762e20ca0871edb75daf670ccacc89b60/src/java/org/lwjgl/opengl/WindowsDisplay.java#L892

So left and right shift keys are handled with `GetAsyncKeyState`, but others are handled by Win32 event system. This explains why simultaneous keys were not able to be processed simultaneously.

phu54321 commented:
Confirmed...

Note the quantized timing graph (bottom right) when playing 180 BPM songs with 60Hz monitor vsynced game.

![20241026_161914_mpc-be64_bDrk9Pclgm](https://github.com/user-attachments/assets/13d3879e-8fce-4ba5-9b39-238f39b3c1fc)


Note that this PR doesn't fixes the problem yet: `BMSPlayerInputProcessor::poll` is still called only 60 times per seconds. It needs more improvement

phu54321 committed:
perf(KeyPressedPreferNative): fix slow polling of fn keys

Calling `Gdx.input.isKeyPressed` acquires lock of global some opengl mutex,
so it takes quite a considerable amount of time.

Filter out excessive calls with GetAsyncKeyState.

phu54321 committed:
perf(KeyPressedPreferNative): Replaced JNA with JNI inside lwjgls.

lwjgl already has implemented multiple JNIs for Windows APIs we need.
Since JNI is much faster than JNA, with a
help of various reflection help we use those
native methods instead of maintaining our own
c++/java JNI build native systems.

- Corollary: removed JNA dependencies

phu54321 commented:
OK everything else is quite optimized, and if I replace the polling thread's `Thread.sleep` to `Thread.yield`. I can easily achieve 8k polling rate. I'm not sure whether this is the right direction.

Example implementation.

```java
Thread polling = new Thread(() -> {
	for (;;) {
		input.poll();
		Thread.yield();
	}
});
```

phu54321 committed:
fix: key setting not working

When we get 'enter' key with `isKeyPressed` variants, `isKeyPressed` may return
`true` value too early before the libgdx event system gets keydown event, so
`lastPressedKey` might not have been updated yet. This makes `Input.Keys.Enter`
(66) to be submitted to `lastPressedKey` on the key input phase. As a result
key input quits immediately due to enter key being pressed.

`lastPressedKey` is ONLY used within the KeyConfiguration class, so this commit
refactors:

- change `lastPressedKey` to `libgdxLastPressedKey` so that it clearly reflects
  that this keycode is fetched from libgdx event system.
- Added a warning that  `lastPressedKey` is *incompatible* with other key fetching
  methods because they fetch keyboard state from different source.
- Modified `KeyConfiguration` so it exclusively uses `libgdxLastPressedKey` for
  key configurations.

phu54321 committed:
perf: cache GetForegroundWindow
